File: 2016 [!DOK]/6.cpp
#include <iostream>
#include <fstream>

using namespace std;

string caesarsCipher(string word, int key) {
    key = key % 26;
    int length = word.length();
    string encrypted = "";

    for (int i = 0; i < length; i++) {
        int newCharNum = (int)(word[i]) + key;
        if (newCharNum > 90) {
            encrypted += (char)(64 + (newCharNum - 90));
        }
        else {
            encrypted += (char)(newCharNum);
        } 
    }

    return encrypted;
}

string decipherCaesarsCipher(string word, int key) {
    key = key % 26;
    key = 26 - key;

    int length = word.length();
    string decrypted = "";

    for (int i = 0; i < length; i++) {
        int newCharNum = (int)(word[i]) + key;
        if (newCharNum > 90) {
            decrypted += (char)(64 + (newCharNum - 90));
        }
        else {
            decrypted += (char)(newCharNum);
        }
    }

    return decrypted;
}

int main()
{
    int const key = 107;
    int const wordsQuantity = 100;
    
    fstream input1, output1;
    fstream input2, output2;
    fstream input3, output3;

    input1.open("dane_6_1.txt", ios::in);
    output1.open("wyniki_6_1.txt", ios::out);

    input2.open("dane_6_2.txt", ios::in);
    output2.open("wyniki_6_2.txt", ios::out);

    input3.open("dane_6_3.txt", ios::in);
    output3.open("wyniki_6_3.txt", ios::out);

    if (input1.good() and output1.good()) {
        for (int i = 0; i < wordsQuantity; i++) {
            string word;

            input1 >> word;

            output1 << caesarsCipher(word, key) << endl;
        }
    }
    else {
        cout << "Problem z plikiem";
    }

    if (input2.good() and output2.good()) {
        for (int i = 0; i < 3000; i++) {
            string word;
            int key;

            input2 >> word >> key;

            output2 << decipherCaesarsCipher(word, key) << endl;
        }
    }
    else {
        cout << "Problem z plikiem";
    }

    if (input3.good() and output3.good()) {
        for (int i = 0; i < 3000; i++) {
            string decoded, encoded;
            bool found = false;

            input3 >> decoded >> encoded;

            for (int i = 0; i < 26; i++) {
                if (caesarsCipher(decoded, i) == encoded) {
                    found = true;
                }
            }

            if (!found) {
                output3 << decoded << endl;
            }
        }
    }
    else {
        cout << "Problem z plikiem";
    }

    input1.close();
    output1.close();

    input2.close();
    output2.close();

    input3.close();
    output3.close();
}

File: 2016/6 - python/6.py
import string


def caesarsCipher(word, key):
    key %= 26
    encrypted = ""
    alphabet = list(string.ascii_uppercase)
    for letter in word:
        letterIndex = alphabet.index(letter)
        shift = (letterIndex + key) % 26
        # print(letterIndex, shift, alphabet[shift])
        encrypted += alphabet[shift]
    return encrypted


def decodeCaesarsCipher(word, key):
    key %= 26
    key = 26 - key
    decrypted = ""
    alphabet = list(string.ascii_uppercase)
    for letter in word:
        letterIndex = alphabet.index(letter)
        shift = (letterIndex + key) % 26
        # print(letterIndex, shift, alphabet[shift])
        decrypted += alphabet[shift]
    return decrypted


input1 = open("dane_6_1.txt", "r")
output1 = open("wyniki_6_1.txt", "w")

input2 = open("dane_6_2.txt", "r")
output2 = open("wyniki_6_2.txt", "w")

input3 = open("dane_6_3.txt", "r")
output3 = open("wyniki_6_3.txt", "w")

# 1
key = 107

words = input1.readlines()
words = list(map(lambda word: word.strip(), words))
print(words)

# print(caesarsCipher("ABCDXYZ", 3))
for word in words:
    encrypted = caesarsCipher(word, key)
    output1.write(encrypted + "\n")


# 2
for i in range(700):
    line = input2.readline().strip()
    word, key = line.split()
    key = int(key)
    # print(word, key)

    decoded = decodeCaesarsCipher(word, key)
    output2.write(decoded + "\n")


# 3
for i in range(3000):
    line = input3.readline().strip()
    w1, w2 = line.split()
    hasCorrectEncryption = False
    print(w1, w2)
    for j in range(26):
        if caesarsCipher(w1, j) == w2:
            hasCorrectEncryption = True
    if not hasCorrectEncryption:
            output3.write(w1 + "\n")

    # decoded = decodeCaesarsCipher(word, key)
    # output2.write(decoded + "\n")


input1.close()
output1.close()

input2.close()
output2.close()

input3.close()
output3.close()
